Support Android data only messages

JBassett opened the pull request.

robbiet480 commented:
LGTM thus far

balloob commented:
This fixes https://github.com/home-assistant/mobile-apps-fcm-push/issues/1 🎉

JBassett commented:
Still need to think about backwards compatibility...  It might be strange for a bit but I think I'm going to version check the app and call the iOS version for backwards compatibility.  But hopefully that only lasts for a couple releases.

balloob commented:
If Android is going to call iOS way of doing it, then let's not call it the "iOS way". That is confusing. 

Also people don't update apps ever, so we will have this support for quite a while I think,

balloob commented on functions/android.js:
```diff
@@ -0,0 +1,42 @@
+module.exports = {
+  createPayload: function createPayload(req) {
+    var payload = {
```
Let's use `let` to avoid surprising behavior of `var`.

balloob commented on functions/index.js:
```diff
@@ -20,141 +23,26 @@ exports.sendPushNotification = functions.https.onRequest(async (req, res) => {
-  if(payload.apns.payload.aps.badge) payload.apns.payload.aps.badge = Number(payload.apns.payload.aps.badge);
-  if(payload.apns.payload.aps.contentAvailable) {
-    payload.apns.headers['apns-push-type'] = 'background';
+    var response = ios.createPayload(req)
```
```suggestion
    [updateRateLimits, response] = ios.createPayload(req)
```

But I would prefer if we use objects. It makes sure the values always map correctly.

In `android.js`: `return { updateRateLimits, payload }`

and here

```suggestion
    { updateRateLimits, payload } = ios.createPayload(req)
```

balloob commented on functions/index.js:
```diff
@@ -20,141 +23,26 @@ exports.sendPushNotification = functions.https.onRequest(async (req, res) => {
+    var response = ios.createPayload(req)
+    updateRateLimits = response[0]
+    payload = response[1]
+  } else if (req.body.registration_info.app_id.indexOf('io.homeassistant.companion.android') > -1) {
```
Should this be `indexOf` or `===` ?

balloob commented on functions/index.js:
```diff
@@ -20,141 +23,26 @@ exports.sendPushNotification = functions.https.onRequest(async (req, res) => {
-    payload.apns.headers['apns-push-type'] = 'alert';
+    console.error('Issue deternmining android vs ios')
+    return res.status(400).send({
+      errorType: 'MissingRegistration',
```
Should be maybe different error type?

balloob commented on functions/index.js:
```diff
@@ -198,6 +86,8 @@ exports.sendPushNotification = functions.https.onRequest(async (req, res) => {
 
   docData.totalCount = docData.totalCount + 1;
 
+  console.log('Sending payload', JSON.stringify(payload));
```
Let's not log in production. That means we would end up logging all notifications.

balloob commented on functions/ios.js:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    if(req.body.data) {
+      if(req.body.data.android) {
```
We can write a simple helper:

```js
for (const key of ['android', 'apns', 'apns_headers']) {
  if (req.body.data[key]) {
    payload[key] = req.body.data[key]
  }
}
```

robbiet480 commented:
> Also people don't update apps ever, so we will have this support for quite a while I think,

They are pretty much forced to update on iOS. The only reason someone would be on an older app is if they didn't upgrade their iOS version.

JBassett commented:
I addressed all comments.  For versioning I figured it would be easiest if we just hard version the notification call.  This should make it easier in the future to make breaking changes.

robbiet480 commented:
LGTM. I can't run it locally because it seems the method of locally testing Firebase Functions is gone, at least for HTTP requests. Will deploy to production and monitor. I think it'll be fine.

robbiet480 commented:
@JBassett This is not good. Errors upon deployment:

```
Error getting document! ReferenceError: ref is not defined
    at handleRequest (/srv/index.js:46:22)
    at exports.sendPushNotification.functions.https.onRequest (/srv/index.js:16:10)
    at cloudFunction (/srv/node_modules/firebase-functions/lib/providers/https.js:49:9)
    at /worker/worker.js:783:7
    at /worker/worker.js:766:11
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)
```
